subsys/retention/retention.c

```diff
@@ -0,0 +1,400 @@
+	int rc;
+
+	if (offset < 0 || ((size_t)offset + size) > (config->size - config->reserved_size)) {
+		/* Disallow reading past the virtual data size or before it */
+		return -EINVAL;
+	}
+
+	retention_lock_take(dev);
+
+	rc = retained_mem_read(config->parent, (config->offset + config->prefix_len +
+			       (size_t)offset), buffer, size);
+
+	retention_lock_release(dev);
+
+	return rc;
+}
+
+int retention_write(const struct device *dev, off_t offset, const uint8_t *buffer, size_t size)
+{
+	const struct retention_config *config = dev->config;
+	struct retention_data *data = dev->data;
+	int rc;
+
+	retention_lock_take(dev);
+
+	if (offset < 0 || ((size_t)offset + size) > (config->size - config->reserved_size)) {
+		/* Disallow reading past the virtual data size or before it */
+		rc = -EINVAL;
+		goto finish;
+	}
+
+	rc = retained_mem_write(config->parent, (config->offset + config->prefix_len +
+				(size_t)offset), buffer, size);
+
+	if (rc < 0) {
+		goto finish;
+	}
+
+	/* Write optional header and footer information, these are done last to ensure data
+	 * validity before marking it as being valid
+	 */
+	if (config->prefix_len != 0 && data->header_written == false) {
+		rc = retained_mem_write(config->parent, config->offset, (void *)config->prefix,
+					config->prefix_len);
+
+		if (rc < 0) {
+			goto finish;
+		}
+
+		data->header_written = true;
+	}
+
+	if (config->checksum != 0) {
+		/* Generating a checksum requires reading out all the data in the region */
+		uint32_t checksum = 0;
+
+		rc = retention_checksum(dev, &checksum);
+
+		if (rc < 0) {
+			goto finish;
+		}
+
+		if (config->checksum == CHECKSUM_TYPE_CRC8) {
+			uint8_t output_checksum = (uint8_t)checksum;
+
+			rc = retained_mem_write(config->parent,
+						(config->offset + config->size - config->checksum),
+						(void *)&output_checksum, sizeof(output_checksum));
+		} else if (config->checksum == CHECKSUM_TYPE_CRC16) {
+			uint16_t output_checksum = (uint16_t)checksum;
+
+			rc = retained_mem_write(config->parent,
+						(config->offset + config->size - config->checksum),
+						(void *)&output_checksum, sizeof(output_checksum));
+		} else if (config->checksum == CHECKSUM_TYPE_CRC32) {
+			rc = retained_mem_write(config->parent,
+						(config->offset + config->size - config->checksum),
+						(void *)&checksum, sizeof(checksum));
+		}
+	}
+
+finish:
+	retention_lock_release(dev);
+
+	return rc;
+}
+
+int retention_clear(const struct device *dev)
+{
+	const struct retention_config *config = dev->config;
+	struct retention_data *data = dev->data;
+	int rc = 0;
+	uint8_t buffer[CONFIG_RETENTION_BUFFER_SIZE];
+	off_t pos = 0;
+
+	memset(buffer, 0, sizeof(buffer));
+
+	retention_lock_take(dev);
+
+	data->header_written = false;
+
+	while (pos < config->size) {
+		rc = retained_mem_write(config->parent, (config->offset + pos), buffer,
+					MIN((config->size - pos), sizeof(buffer)));
+
+		if (rc < 0) {
+			goto finish;
+		}
+
+		pos += sizeof(buffer);
```

Comment:
Shouldn't this be 
```
pos += MIN((config->size - pos), sizeof(buffer));
```
?

Reply:
I don't think it should cause a bug though

Reply:
Fixed